Increase default OIDC session-age-extension to minimize chances of dropping expired ID tokens (#41745)
quarkusio/quarkus: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc

To use the refresh token, you should carefully configure the session cookie age.
The session age should be longer than the ID token lifespan and close to or equal to the refresh token lifespan.

You calculate the session cookie age by adding the lifespan value of the current ID token and the values of the `quarkus.oidc.authentication.session-age-extension` and `quarkus.oidc.token.lifespan-grace` properties.

[TIP]
====
You use the `quarkus.oidc.token.lifespan-grace` property only to consider some small clock skews.
====

[TIP]
====
You use the `quarkus.oidc.authentication.session-age-extension` property to significantly extend the session lifespan in order to

> make even expired ID token available to the application.

Author: I'm not sure what this suggestion proposes ?

> make even expired ID token available to the application.

Reviewer: I Just applied the one sentence-per line rule. The text I am proposing to delete here was moved to a previous row.
Let me double-check that.

> make even expired ID token available to the application.

Reviewer: Yeah, I reformulate it so that it is good to use on a higher row. Then, I deleted the portion of the original text that was no longer needed.
![image](https://github.com/user-attachments/assets/66ed2ba3-f47a-4b0b-97cc-3ee6baaa6d1b)

> make even expired ID token available to the application.

Reviewer: The text from the line 1227 is on the row 1226 now.

However, if `quarkus.oidc.authentication.session-age-extension` is set to a reasonable idle-time duration of several hours,
longer than an ID token life-span, then, even if the ID token has expired, it can still be made available to the application which may refresh it
or redirect the user to a session expired page, because the session cookie which keeps ID token is still available during the current authenticated user's request.

For example, let's assume you expect a typical user activity to last for 30 minutes.
Therefore, you have configured OIDC provider to issue ID tokens which are valid for 30 minutes only.
But, you would also like to have a user's ID token refreshed and the session renewed after this user has returned from a one hour break.
In this case, set this property to 90 minutes, otherwise, Quarkus will not be able to detect a session cookie, and the user will be required to re-authenticate.

An important point is that extending the session cookie's age with `quarkus.oidc.authentication.session-age-extension`
is only about maximizing the chances of making even an expired ID token associated with this cookie available to the application for further processing.

> It improves a user experience, otherwise, an authenticated user may be surprised by seeing an immediate re-authentication challenge

Reviewer: ```suggestion
It improves a user experience; otherwise, an authenticated user may be surprised to see an immediate re-authentication challenge if the ID token and the session cookie that holds it have expired.
```

> if the ID token has expired together with the session cookie which holds it.

Reviewer: ```suggestion
```

> By default, this property will be set to 1 hour if either `quarkus.oidc.token.refresh-expired` or `quarkus.oidc.authentication.session-expired-page`

Reviewer: ```suggestion
By default, this property will be set to 1 hour if either `quarkus.oidc.token.refresh-expired` or `quarkus.oidc.authentication.session-expired-page` features are enabled since they can only be effective if an expired ID token is available.
```

> features are enabled since they can only be effective if an expired ID token is available.

Reviewer: ```suggestion
```

For example, expect a typical user activity to last 30 minutes.
Therefore, you have configured the OIDC provider to issue ID tokens, which are valid for 30 minutes only.
However, you would also like to refresh a user's ID token and renew the session after this user returns from a one-hour break.
In this case, set this property to 90 minutes; otherwise, Quarkus will not be able to detect a session cookie, and the user will be required to re-authenticate.

The important point is that extending the session cookie's age with `quarkus.oidc.authentication.session-age-extension` maximizes the chances of making an expired ID token available to the application.
This token, associated with the cookie, can then be used for further processing.

It improves a user experience; otherwise, an authenticated user may be surprised to see an immediate re-authentication challenge if the ID token and the session cookie that holds it have expired.

By default, this property will be set to 1 hour if either `quarkus.oidc.token.refresh-expired` or `quarkus.oidc.authentication.session-expired-page` features are enabled since they can only be effective if an expired ID token is available.

> ====

Reviewer 2: If you'd like a more concise version of this tip, here’s a refined option. Feel free to use any parts that best fit your needs.

```
Use the `quarkus.oidc.authentication.session-age-extension` property to extend the session lifespan and ensure that the application can access expired ID tokens when needed.

By default, the session cookie's `max-age` property matches the ID token lifespan. When the session cookie expires and the user attempts to access the application again, the authenticated user is redirected to the OpenID Connect (OIDC) provider for re-authentication.

To improve session continuity, set `quarkus.oidc.authentication.session-age-extension` to a reasonable idle-time duration longer than the ID token lifespan, such as several hours. This setting ensures that an expired ID token remains available, allowing the application to refresh the token or redirect the user to a "session expired" page while the session cookie persists during the authenticated user's request.

For example, if typical user activity lasts 30 minutes, you might configure the OIDC provider to issue ID tokens valid for the same duration. To allow session renewal when a user returns after a one-hour break, set the property to 90 minutes. Otherwise, the application cannot detect a session cookie, and the user must re-authenticate.

Extending the session cookie's age with `quarkus.oidc.authentication.session-age-extension` increases the likelihood that an expired ID token remains available for further processing. This improves the user experience by reducing unexpected re-authentication prompts.

By default, if either `quarkus.oidc.token.refresh-expired` or `quarkus.oidc.authentication.session-expired-page` is enabled, this property is set to one hour to ensure these features function correctly.
```
